egs/librispeech/asr/simple_v1/train_fast.py

```diff
@@ -0,0 +1,286 @@
+    print(dense_fsa_vec.dense_fsa_vec)
+    target_graph = k2.intersect_dense_pruned(decoding_graph, dense_fsa_vec, 10,
+                                             10000, 0)
+    tot_scores = -k2.get_tot_scores(target_graph, True, False).sum()
+    if training:
+        optimizer.zero_grad()
+        tot_scores.backward()
+        clip_grad_value_(model.parameters(), 5.0)
+        optimizer.step()
+
+    objf = tot_scores.detach().cpu()
+    total_objf = objf.item()
+    total_frames = nnet_output.shape[0]
+
+    return total_objf, total_frames
+
+
+def get_validation_objf(dataloader, model, device, graph, symbols):
+    total_objf = 0.
+    total_frames = 0.  # for display only
+
+    model.eval()
+
+    for batch_idx, batch in enumerate(dataloader):
+        objf, frames = get_objf(batch, model, device, graph, symbols, False)
+        total_objf += objf
+        total_frames += frames
+
+    return total_objf, total_frames
+
+
+def train_one_epoch(dataloader, valid_dataloader, model, device, graph,
+                    symbols, optimizer, current_epoch, num_epochs):
+    total_objf = 0.
+    total_frames = 0.
+
+    model.train()
+    for batch_idx, batch in enumerate(dataloader):
+        curr_batch_objf, curr_batch_frames = get_objf(batch, model, device,
+                                                      graph, symbols, True,
+                                                      optimizer)
+
+        total_objf += curr_batch_objf
+        total_frames += curr_batch_frames
+
+        if batch_idx % 100 == 0:
+            logging.info(
+                'processing batch {}, current epoch is {}/{} '
+                'global average objf: {:.6f} over {} '
+                'frames, current batch average objf: {:.6f} over {} frames'.
+                format(
+                    batch_idx,
+                    current_epoch,
+                    num_epochs,
+                    total_objf / total_frames,
+                    total_frames,
+                    curr_batch_objf / curr_batch_frames,
+                    curr_batch_frames,
+                ))
+
+        if valid_dataloader and batch_idx % 1000 == 0:
+            total_valid_objf, total_valid_frames = get_validation_objf(
+                dataloader=valid_dataloader,
+                model=model,
+                device=device,
+                graph=graph,
+                symbols=symbols)
+            model.train()
+            logging.info(
+                'Validation average objf: {:.6f} over {} frames'.format(
+                    total_valid_objf / total_valid_frames, total_valid_frames))
+    return total_objf
+
+
+def main():
+    # load L, G, symbol_table
+    lang_dir = 'data/lang_nosp'
+    with open(lang_dir + '/L.fst.txt') as f:
+        L = k2.Fsa.from_openfst(f.read(), acceptor=False)
+
+    with open(lang_dir + '/G.fsa.txt') as f:
+        G = k2.Fsa.from_openfst(f.read(), acceptor=True)
+
+    with open(lang_dir + '/words.txt') as f:
```

Comment:
`k2.SymbolTable` can be constructed directly from a filename. Please refer to

https://github.com/k2-fsa/k2/blob/master/k2/python/k2/symbol_table.py#L69

```python
    @staticmethod
    def from_file(filename: str) -> 'SymbolTable':
        '''Build a symbol table from file.
```

Reply:
Are the main differences between `from_file()/from_str()` and `from_openfst()` lie in number of final states allowed, label to the final state (-1 or not), and whether final weight is allowed?

Reply:
`from_file/from_str` is for k2 format, for FSA, the input format is
```
src_state dest_state label score
....
final_state
```
`from_openfst` is for openfst format, for FSA, the input format is
```
src_state dest_state lable score
...
final_state final_score
```
(for FST, each line would be `src_state dest_state lable aux_label score`, see here for details https://github.com/k2-fsa/k2/blob/d96e93c854e36e1b92fd6b0eb0f1b3789f59776c/k2/python/k2/fsa.py#L635-L718)

@freewym, in k2, each FSA  have only one final state which has no weights on it, all arcs entring this final state (we call it final-arcs) must have `label == -1`. So for openfst format whose final_state has a weight on it, we will create a super final state, and then the original final_state in openfst will have a arc entering to the super final state, with `label==-1` and `weight=final_score`. See here for details https://github.com/k2-fsa/k2/blob/d96e93c854e36e1b92fd6b0eb0f1b3789f59776c/k2/csrc/fsa_utils.cu#L314-L319

Reply:
> `k2.SymbolTable` can be constructed directly from a filename. Please refer to
> 
> https://github.com/k2-fsa/k2/blob/master/k2/python/k2/symbol_table.py#L69
> 
> ```python
>     @staticmethod
>     def from_file(filename: str) -> 'SymbolTable':
>         '''Build a symbol table from file.
> ```

Updated